### PR TITLE
Gui improvement

### PR DESCRIPTION
**Pull request by modyf01**

### CODE REVIEW

File: midi_app_controller/gui/utils.py [hunk 1]
```diff
@@ -141,8 +141,8 @@ def get_selected_action_id(self) -> Optional[str]:
 
 def is_subpath(path: Path, subpath: Path) -> bool:
     """Checks if one path represents a file/directory inside the other directory."""
-    path_str = str(subpath.resolve().absolute())
-    subpath_str = str(path.resolve().absolute())
+    path_str = str(path.resolve().absolute())
+    subpath_str = str(subpath.resolve().absolute())
     return path_str.startswith(subpath_str)
```
Comment (lukaszsmolinski) [inline midi_app_controller/gui/utils.py:146]: Everywhere it is used, the `subpath` is the first argument. So they need to be switched here.

File: midi_app_controller/gui/binds_editor.py [hunk 2]
```diff
@@ -364,85 +378,92 @@ def __init__(
         """
         super().__init__()
 
+        self.setWindowTitle("Edit Binds")
         self.binds = binds.copy(deep=True)
         self.save_binds = save_binds
 
+        name_label = QLabel("Name:")
```
Comment (lukaszsmolinski) [inline midi_app_controller/gui/binds_editor.py:385]: `name_label` is not used anywhere else so maybe just `name_layout.addWidget(QLabel("Name:"))`. It's done this way in the rest of the code.

File: midi_app_controller/gui/binds_editor.py [hunk 3]
```diff
@@ -364,85 +378,92 @@ def __init__(
...
         self.name_edit = QLineEdit(binds.name)
+        name_layout = QHBoxLayout()
+        name_layout.addWidget(name_label)
+        name_layout.addWidget(self.name_edit)
+
+        self.show_action_names_checkbox_button = QCheckBox("Show action names")
+        self.show_action_names_checkbox_button.setChecked(False)
```
Comment (lukaszsmolinski) [inline midi_app_controller/gui/binds_editor.py:392]:
```suggestion
        self.show_action_names_checkbox_button.setChecked(True)
```

File: midi_app_controller/gui/binds_editor.py [hunk 4]
```diff
@@ -364,85 +378,92 @@ def __init__(
...
+        self.show_action_names_checkbox_button.stateChanged.connect(
+            self._toggle_names_mode
+        )
+
+        self.show_action_names_checkbox_knob = QCheckBox("Show action names")
+        self.show_action_names_checkbox_knob.setChecked(False)
```
Comment (lukaszsmolinski) [inline midi_app_controller/gui/binds_editor.py:398]:
```suggestion
        self.show_action_names_checkbox_knob.setChecked(True)
```

File: midi_app_controller/gui/binds_editor.py [hunk 5]
```diff
@@ -364,85 +378,92 @@ def __init__(
...
+        )
 
         # Save/exit buttons.
-        toggle_names_mode_button = QPushButton("Toggle names mode")
-        toggle_names_mode_button.clicked.connect(self._toggle_names_mode)
         save_and_exit_button = QPushButton("Save and exit")
         save_and_exit_button.clicked.connect(self._save_and_exit)
         exit_button = QPushButton("Exit")
-        exit_button.clicked.connect(self._exit)
+        exit_button.clicked.connect(self.close)
```
Comment (lukaszsmolinski) [inline midi_app_controller/gui/binds_editor.py:407]:
```suggestion
        exit_button.clicked.connect(self._exit)
```
There is (not on main yet) some code in `self._exit` that waits for threads etc.

File: midi_app_controller/gui/binds_editor.py [hunk 6]
```diff
@@ -364,85 +378,92 @@ def __init__(
...
+
+        save_and_exit_button.setFixedWidth(button_width)
+        exit_button.setFixedWidth(button_width)
 
         buttons_layout = QHBoxLayout()
-        buttons_layout.addWidget(toggle_names_mode_button)
+        spacer = QSpacerItem(40, 20, QSizePolicy.Expanding, QSizePolicy.Minimum)
+        buttons_layout.addItem(spacer)
         buttons_layout.addWidget(save_and_exit_button)
         buttons_layout.addWidget(exit_button)
```
Comment (lukaszsmolinski) [inline midi_app_controller/gui/binds_editor.py:419]: The buttons were in different order on the visualization.

File: midi_app_controller/gui/binds_editor.py [hunk 7]
```diff
@@ -462,6 +483,17 @@ def _wait_for_worker_threads(self):
         for thread in self.knobs_widget.thread_list:
             thread.wait()
 
+    def updateCheckBox(self):
```
Comment (lukaszsmolinski) [inline midi_app_controller/gui/binds_editor.py:486]: snake_case should be used for method names. And maybe something more verbose like `update_action_names_checkboxes`?

**Review by lukaszsmolinski — CHANGES_REQUESTED** [review]

**Review by lukaszsmolinski — CHANGES_REQUESTED** [review]
Fix the title of this PR.

File: midi_app_controller/gui/utils.py [hunk 8]
```diff
@@ -141,9 +141,9 @@ def get_selected_action_id(self) -> Optional[str]:
 
 def is_subpath(path: Path, subpath: Path) -> bool:
     """Checks if one path represents a file/directory inside the other directory."""
-    path_str = str(subpath.resolve().absolute())
-    subpath_str = str(path.resolve().absolute())
-    return path_str.startswith(subpath_str)
+    path_str = str(path.resolve().absolute())
+    subpath_str = str(subpath.resolve().absolute())
+    return subpath_str.startswith(path_str)
```
Comment (Aleshkev) [inline midi_app_controller/gui/utils.py:146]: Regarding comment by @lukaszsmolinski, I think this version is correct and when I first wrote the function I meant it like this, with `path = A/B, subpath = A/B/C -> True`, so it is called everywhere correctly. I'm not sure why the variables were flipped in the function body. 

However, it has come to my attention "subpath" is ambiguous and can mean opposite things in this context. I'd change the docstring to ``"Checks if `subpath` represents a file/directory inside the directory represented by `path`"``.

Comment (lukaszsmolinski) [reply midi_app_controller/gui/utils.py:146]: Yes, now it works. When I reviewed it, the last line was `path_str.startswith(subpath_str)`.

File: midi_app_controller/models/binds.py [hunk 9]
```diff
@@ -1,5 +1,6 @@
 from typing import Optional
-from pydantic import BaseModel, Field, root_validator
+from pydantic import BaseModel, Field
+from pydantic_compat import model_validator
```
Comment (lukaszsmolinski) [inline midi_app_controller/models/binds.py:3]: It's for `v1` of `pydantic`. In `v2` the `model_validator` should be imported from `pydantic`.

File: midi_app_controller/state/state_manager.py [hunk 10]
```diff
@@ -242,14 +242,22 @@ def load_state(self):
         # want to use them. So we make sure that all file paths are ok for us
         # to use.
         controller_file = state.selected_controller_path
-        if not any(is_subpath(d, controller_file) for d in Config.CONTROLLER_DIRS):
+        if controller_file is None:
+            return
+
+        if not any(
+            d and is_subpath(d, controller_file) for d in Config.CONTROLLER_DIRS
```
Comment (lukaszsmolinski) [inline midi_app_controller/state/state_manager.py:249]: We don't expect to store `None` in `Config.CONTROLLER_DIRS`. The `d and` is really not needed. I would rather get an exception if this happens.

File: requirements-test.txt [hunk 11]
```diff
@@ -6,7 +6,7 @@ mypy-extensions==1.0.0
 packaging==23.2
 pluggy==1.3.0
 psygnal==0.9.5
-pydantic==1.10.13
+pydantic>=2.0.0
```
Comment (lukaszsmolinski) [inline requirements-test.txt:9]: Exact versions are used in this file. And you forgot about updating the version in `setup.cfg`.

**Review by lukaszsmolinski — CHANGES_REQUESTED** [review]

File: midi_app_controller/actions/_tests/test_bound_controller.py [hunk 12]
```diff
@@ -70,18 +70,24 @@ def bound_controller(binds, controller, actions) -> BoundController:
...
+            if hasattr(data[key], key2):
+                assert getattr(data[key], key2) == x[key][key2]
+
+
 def test_button_actions(actions):
     data = {"action_press": actions[0]}
-    actions = ButtonActions(**data)
-
-    assert actions.dict() == data
+    validate_actions(data, ButtonActions)
```
Comment (lukaszsmolinski) [inline midi_app_controller/actions/_tests/test_bound_controller.py:85]: Why not just `actions.model_dump() == data` (where `actions = ButtonActions(**data)`)? It's done this way in `test_utils.py` and `test_controller.py`.

File: midi_app_controller/gui/binds_editor.py [hunk 13]
```diff
@@ -1,6 +1,7 @@
 # TODO Move style somewhere else in the future to make this class independent from napari.
 from typing import Callable, Optional
 
+from PyQt5.QtWidgets import QTabWidget, QCheckBox, QSpacerItem, QSizePolicy
```
Comment (lukaszsmolinski) [inline midi_app_controller/gui/binds_editor.py:4]: We use `qtpy`.

**Review by lukaszsmolinski — CHANGES_REQUESTED** [review]

File: midi_app_controller/actions/_tests/test_bound_controller.py (hunk 12 above)
Comment (modyf01) [reply midi_app_controller/actions/_tests/test_bound_controller.py:85]: data = {'action_press': Action(id='Action1', title='sdfgsdfg', category=None, tooltip=None, status_tip=None, icon=None, icon_visible_in_menu=True, enablement=None, short_title=None, toggled=None, callback=<function actions.<locals>.<lambda> at 0x7a1b7b047e20>, menus=None, keybindings=None, palette=True)}

actions.model_dump() = {'action_press': {'category': None, 'enablement': None, 'icon': None, 'icon_visible_in_menu': True, 'id': 'Action1', 'short_title': None, 'status_tip': None, 'title': 'sdfgsdfg', 'toggled': None, 'tooltip': None}}

**Review by lukaszsmolinski — APPROVED** [review]

**Review by Aleshkev — APPROVED** [review]

**Pull request merged**